Response hierarchy - Closes LIBCLOUD-97

caio opened the pull request:
Hello,

This patchset implements two base response classes XmlResponse and JsonResponse in order to reduce code duplication and adapts the common and specialized drivers to use the new classes.

Tests passing for each commit separately.

caio committed:
Implement XmlResponse and JsonResponse base classes

To avoid repeating a common pattern of parsing Xml- and Json-based
responses, create a base class that does the dirty work so the drivers
may simply use them.

Ref: https://issues.apache.org/jira/browse/LIBCLOUD-97

caio committed:
Adapt common drivers to use the new {Xml,Json}Response

caio committed:
Adapt specialized drivers to use {Xml,Json}Response

Kami commented on libcloud/compute/drivers/bluebox.py:
```diff
@@ -83,12 +79,12 @@ NODE_STATE_MAP = { 'queued': NodeState.PENDING,
 
-class BlueboxResponse(Response):
+class BlueboxResponse(JsonResponse):
     def parse_body(self):
```
Looks like it's safe to remove `parse_body` here.

`parse_error` doesn't call `parse_body` here and `parse_body` always gets called after `parse_error` if `success()` returns `True`. Also looking at the code all API calls seem to return JSON.

Kami commented on libcloud/compute/drivers/brightbox.py:
```diff
@@ -31,18 +31,18 @@ except ImportError:
 
     def parse_body(self):
         if self.headers['content-type'].split('; ')[0] == 'application/json' and len(self.body) > 0:
-            return json.loads(self.body)
```
Hm, thinking if we should add `parse_zero_length_body` attribute to the base XMLResponse & JSONResponse class.

Then we could so something like this there:

``` python
if len(self.body) == 0 and not self.parse_zero_length_body:
  return self body
...
```

caio replied:
In this case in particular (brightbox driver) this wouldn't help much since there's still the need to check for the content-type, right?

This makes me wonder if we shouldn't move this content-type verification into the {Json,Xml}Response classes and create a behavior toggle just like this `parse_zero_length_body` to let the specialized response class decide what to do if the header is not as expected (`ignore_content_type` to allow parsing regardless of the content-type sent back).

This is probably not the best place for brainstorming, though :)

Kami commented on libcloud/compute/drivers/elasticstack.py:
```diff
@@ -121,15 +121,7 @@ class ElasticStackResponse(Response):
     def parse_body(self):
         if not self.body:
```
Yeah, if we add `parse_zero_length_body` attribute we could also get rid of overriding `parse_body` here...

Kami commented on libcloud/compute/drivers/slicehost.py:
```diff
@@ -35,26 +34,13 @@ class SlicehostResponse(Response):
         # application/xml" header. booya.
         if not self.body or len(self.body) <= 1:
```
Same here with regards to the `parse_zero_length_body`attribute.

caio replied:
The `len(self.body) == 0` would fail here, should we change the `parse_zero_length_body` logic to test the stripped body? I.e.: 

``` python
if len(self.body.strip()) == 0 and not self.parse_zero_length_body:
    return self body
```

Or maybe we could strip it directly on `Response.__init__`

Kami replied:
Yeah, I would be OK with something like that in the `Response` class.

caio replied:
Done that in the two new commits. More lines removed :)

Kami committed:
OpenStack and Opsource driver style and code cleanup. No functional changes.
Contributed by Hutson Betts <hut606 at gmail dot com> part of LIBCLOUD-123.


git-svn-id: https://svn.apache.org/repos/asf/libcloud/trunk@1187872 13f79535-47bb-0310-9956-ffa450edef68

caio committed:
Don't try to parse an empty body on {Xml,Json}Response classes

caio committed:
Update specialized drivers due to new parse_zero_length_body logic

Kami committed:
Add OpenNebulaNetwork class and ex_list_networks() method to the OpenNebula
driver. Also be consistent and make node states constants uppercase and clean up
the style (double quotes to single quotes, etc.).

Contributed by Hutson Betts <hut606 at gmail dot com> and part of LIBCLOUD-121.


git-svn-id: https://svn.apache.org/repos/asf/libcloud/trunk@1187946 13f79535-47bb-0310-9956-ffa450edef68

caio commented:
Pull request merged as single blob: https://github.com/apache/libcloud/commit/211a6b5219a94f0d11e42e5198f7729cc361cb2f
Thanks for the reviews!
